### DISCUSSION TITLE
Fix index page handling for build format `file`

### DISCUSSION THREAD

**Discussion by shishkin** [withastro/roadmap#474] [Proposal]
### Body

# Summary

I suggest a breaking change for 3.0.0 to change output file path of `index.astro` (and other supported file extensions) pages with build format `file`.

# Background & Motivation

Follow up on https://github.com/withastro/astro/issues/6167:

I have a configuration with [format file](https://docs.astro.build/en/reference/configuration-reference/#buildformat) and following pages:

```
foo/index.md
foo/other.md
index.md
```

The build output is confusing:

```
▶ src/pages/index.md
  └─ /index.html                  ✅ This is expected
▶ src/pages/foo/index.md
  └─ /foo.html                    🤯 WTF?
▶ src/pages/foo/other.md
  └─ /foo/other.html              ✅ This is expected
```

How I understood the file format setting is that Astro would not add artificial index.html but I didn't expect it to actively remove explicitly created index pages, rename them and move them one level up.

Apart from the confusion of URLs, it actively breaks relative links in the pages as they are not siblings anymore.

I understand that changing this behavior would be a breaking change. Maybe a way to enforce keeping index.html would be a better fix. I tried `foo/index.html.md` but got `foo.html.html`, which is even more confusing.

Using build format `file` is required for SSG environments with limited HTTP server support. For example GitLab CI job artifact browsing endpoint (not to be confused with GitLab Pages) didn't resolve index files in my testing.

# Goals

- Fix unexpected breakages in relative URLs
- Make handling of pages consistent and predictable

**Comment by candeynasa** [comment]
We have a hierarchy of directories and each needs to have its own index.html page to automatically display when entering that directory. This may be a show stopper for us, since we need to preserve the existing URLs that are referenced in publications.

**Comment by ematipico** [comment]
This is probably related to https://github.com/withastro/astro/issues/6947
